packages/mui-styled-engine/src/index.d.ts

```diff
@@ -58,20 +59,24 @@ export interface CSSOthersObjectForCSSObject {
   [propertiesName: string]: CSSInterpolation;
 }
 
-// Omit variants as a key, because we have a special handling for it
-export interface CSSObject
-  extends CSSPropertiesWithMultiValues,
-    CSSPseudos,
-    Omit<CSSOthersObject, 'variants'> {}
-
-interface CSSObjectWithVariants<Props> extends Omit<CSSObject, 'variants'> {
-  variants: Array<{
-    props: Props | ((props: Props) => boolean);
-    style:
-      | CSSObject
-      | ((args: Props extends { theme: any } ? { theme: Props['theme'] } : any) => CSSObject);
-  }>;
-}
+export interface CSSObject extends CSSPropertiesWithMultiValues, CSSPseudos, CSSOthersObject {}
+
+export type CSSObjectWithVariants<Props> = CSSPropertiesWithMultiValues &
+  CSSPseudos & {
+    variants: Array<{
+      props:
+        | Partial<Props>
+        | ((
+            props: Partial<Props> & {
+              ownerState: Partial<Props>;
+            },
+          ) => boolean);
+      style:
+        | CSSObject
+        | ((args: Props extends { theme: any } ? { theme: Props['theme'] } : any) => CSSObject);
+    }>;
+    [propertiesName: string]: unknown | CSSInterpolation;
```

Comment:
We can't extend `CSSOthersObject` without introducing a `variants: unknown` (`Omit<CSSOthersObject, 'variants'>` doesn't work, as explained in the description. That's why we add `[propertiesName: string]: unknown | CSSInterpolation;` manually